Fix remote and realm join

Yehonal opened the pull request:
This fixes connection when server is not in same host of wowser

removing localhost hardcoded inside and implementing port configuration.

Also sending correct opcode when realm is not 0 

app.writeUnsignedInt(this.realm.id);

Commit:
Merge branch 'master' of https://github.com/wowserhq/wowser

Commit:
fixed multiple realms selection

Commit:
Merge branch 'master' of git://github.com/wowserhq/wowser.git into fix-remote

Commit:
Merge branch 'master' of git://github.com/wowserhq/wowser.git into fix-remote

timkurvers commented on src/lib/game/handler.js:
```diff
@@ -27,9 +27,10 @@ class GameHandler extends Socket {
   // Connects to given host through given port
-  connect(host, port) {
+  connect(host, realm) {
+    this.realm=realm;
```
Could you add spaces around operators, adhering to the [styleguide](http://eslint.org/docs/rules/space-infix-ops)?

Yehonal replied:
I've fixed it, however i suggest you to make a "confirm and squash" before accept this PR

https://help.github.com/articles/about-pull-request-merge-squashing/

timkurvers commented on src/lib/game/handler.js:
```diff
@@ -27,9 +27,10 @@ class GameHandler extends Socket {
   }
 
   // Connects to given host through given port
```
Could you update this comment to reflect the change in method signature?

timkurvers commented on src/components/realms/index.jsx:
```diff
@@ -32,7 +32,7 @@ class RealmsScreen extends React.Component {
 
   connect(realm) {
-    session.game.connect('localhost', realm.port);
+    session.game.connect(session.authstate.host, realm);
```
Is this state necessary? Wouldn't `session.auth.host` suffice?

timkurvers commented on src/components/realms/index.jsx:
```diff
@@ -44,7 +44,11 @@ class RealmsScreen extends React.Component {
+    const realms = session.realms.list;
+    this.setState({
+        realm: realms[event.target.selectedIndex],
+        realms: realms
```
Where is `realms` used?

Commit:
code style

Commit:
Merge branch 'fix-remote' of git@github.com:udw/wowser.git into fix-remote

Commit:
some cleaning

Yehonal commented:
I've cleaned it 

However about auth state, i don't remember why i needed to create that state... but there are only 2 possibilities:

1) i've not noticed the session.auth.host

2) session.auth.host is not the same of component state ( value inserted into text input ) . I've checked it reading the code, but seems that are the same values. I cannot test it now

However i've cleaned all the code and i think you can squash and merge if you want.

timkurvers commented:
Had you been able to test whether the `setState` changes are required?

fallenoak commented:
Since Wowser currently only runs in the browser, its ability to connect to a given host and port via TCP is limited to what's exposed by a Websockify proxy. See here: https://github.com/wowserhq/wowser#socket-proxies

While supporting arbitrary hostname and port combinations is something we'd like to do, I'm not really sure there's much to be gained by adding this before Wowser is either capable of running in another context (like an Electron-ified desktop app), or is capable of specifying configuration options for the proxy server.

timkurvers commented:
Great catch @fallenoak, had completely forgotten about that!

Yehonal commented:
I'm using it with your proxy configured to handle this connection. I've
tested in any kind of configurations: wowser and proxy on same host of
server in a local context , wowser and proxy in remote on same host of
server, wowser and proxy on remote or on local with server in another host.
With chrome or firefox I've any kind of problem. This configuration is
needed to allow to host wowser on remote and allowing people to connect
without any proxy configured..with a small trick on server proxy.

If you explain me where is exaxtly the problem i can test it better

Il 01/Ago/2016 14:42, "Tim Kurvers" <notifications@github.com> ha scritto:

> Great catch @fallenoak <https://github.com/fallenoak>, had completely
> forgotten about that!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wowserhq/wowser/pull/165#issuecomment-236569853>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAI-lPrc5uCQPskabHltAjH6jZr3G0buks5qbemjgaJpZM4JV-4P>
> .
>

fallenoak commented:
@Yehonal Right, but if the auth server is returning some specific port number, and that port is actually running a Websockify proxy, how is the real World of Warcraft client still able to connect?
